File: src/goserver/nf/game/medicine.go
/**
 * Created by nazarigonzalez on 29/10/16.
 */

package game

import (
	"../utils"
	"gopkg.in/nazariglez/go-rbush.v0"
	"math"
	"math/rand"
)

const (
	rad180 = math.Pi
	rad    = 2 * math.Pi

	medicineRadarTime      = 1
	medicineSpeedMin       = 50
	medicineSpeedMax       = 130
	medicineTimeToRespawn  = 90
	medicineRadius         = 75
	medicineSearchDistance = 0.25
)

type Medicine struct {
	State EntityState

	Id int

	Speed, Angle, Radius, AngleDir float64

	Position, vectorSpeed *Point

	changeMovementTime        float64
	Target                    *Actor
	targetRadarTime, distance float64
	dirty                     bool

	Box                       *rbush.Box
	MapSize, speedMax         float64
	TimeToRespawn, SearchTime float64

	Data *MedicineData
}

func NewEntityMedicine() *Medicine {
	box := rbush.Box{}

	return &Medicine{
		Id:       -1,
		State:    STATE_BORN,
		Box:      &box,
		Position: &Point{},
		Radius:   medicineRadius,

		Data: &MedicineData{},

		vectorSpeed: &Point{},
	}
}

func (m *Medicine) Initialize(id, mapSize int) {
	m.Id = id
	m.Box.Data = id
	m.MapSize = float64(mapSize)

	m.distance = m.MapSize * medicineSearchDistance
	m.speedMax = medicineSpeedMax + (rand.Float64() * 30)
	m.setRandomParams()
}

func (m *Medicine) Reset() {
	m.Id = -1
	m.MapSize = 0
	m.Radius = medicineRadius
	m.Speed = 0
	m.Angle = 0
	m.AngleDir = 0

	if m.Target != nil {
		m.Target.FindingByMedicine = -1
	}

	m.Target = nil
	m.changeMovementTime = 0
	m.targetRadarTime = 0
	m.dirty = false
	m.distance = 0
	m.TimeToRespawn = 0
	m.SearchTime = 0

	m.vectorSpeed.Set(0, 0)
	m.Position.Set(0, 0)
}

func (m *Medicine) Die() {
	m.State = STATE_DEAD
	m.TimeToRespawn = medicineTimeToRespawn + rand.Float64()*30
}

func (m *Medicine) Update(delta float64) {
	if m.State == STATE_DEAD {
		m.TimeToRespawn -= delta
		return
	}

	m.Position.X += m.vectorSpeed.X * delta
	m.Position.Y += m.vectorSpeed.Y * delta

	if m.Target != nil {
		if m.Target.State == STATE_DEAD || m.Target.Crazy {
			m.Target.FindingByMedicine = -1
			m.Target = nil
			m.setRandomParams()
		} else {
			m.targetRadarTime -= delta

			if m.targetRadarTime <= 0 {
				a := m.Position.X - m.Target.Position.X
				b := m.Position.Y - m.Target.Position.Y

				distance := math.Abs(math.Sqrt(a*a + b*b))

				if distance > m.distance+m.Target.Radius {
					m.Target.FindingByMedicine = -1
					m.Target = nil
					m.setRandomParams()
				} else {
					m.AngleDir = math.Atan2(m.Target.Position.Y-m.Position.Y, m.Target.Position.X-m.Position.X)
				}
			}

		}
	} else {
		m.changeMovementTime -= delta

		if m.changeMovementTime <= 0 {
			m.setRandomParams()
		}
	}

	if m.AngleDir != m.Angle {
		if m.Angle < m.AngleDir {
			m.Angle = math.Mod(m.Angle+rad180*delta, rad)

			if m.Angle < m.AngleDir {
				m.Angle = m.AngleDir
			}
		} else {
			m.Angle = math.Mod(m.Angle-rad180*delta, rad)

			if m.Angle < m.AngleDir {
				m.Angle = m.AngleDir
			}
		}
		m.dirty = true
	}

	if m.dirty {
		m.calcSpeed()
	}

}

func (m *Medicine) UpdateBox() {
	m.Box.MinX = m.Position.X - m.Radius
	m.Box.MinY = m.Position.Y - m.Radius
	m.Box.MaxX = m.Position.X + m.Radius
	m.Box.MaxY = m.Position.Y + m.Radius
}

func (m *Medicine) setRandomParams() {
	if m.State == STATE_BORN {
		m.State = STATE_ALIVE
	}

	m.AngleDir = math.Mod(math.Floor(rand.Float64()*360)*utils.DEG_TO_RAD, rad)
	m.Speed = medicineSpeedMin
	m.changeMovementTime = 3 + rand.Float64()*6
	m.calcSpeed()
}

func (m *Medicine) SetTarget(a *Actor) {
	if m.Target != nil {
		m.Target.FindingByMedicine = -1
	}

	m.targetRadarTime = medicineRadarTime
	m.Target = a
	m.Target.FindingByMedicine = m.Id
	m.Speed = medicineSpeedMax
	m.AngleDir = math.Atan2(m.Target.Position.Y-m.Position.Y, m.Target.Position.X-m.Position.X)
	m.calcSpeed()
}

func (m *Medicine) IsBestTarget(a *Actor) bool {
	if m.Target == nil || m.Target.State == STATE_DEAD || m.Target.Crazy { //todo actor.crazy, why?? it's a bug?
		return true
	}

	return a.Mass > m.Target.Mass
}

func (m *Medicine) FixPosition(delta, x, y float64) {
	change := false

	if m.Position.X < -x {
		m.Position.X = -x
		change = true
	} else if m.Position.X > x {
		m.Position.X = x
		change = true
	}

	if m.Position.Y < -y {
		m.Position.Y = -y
		change = true
	} else if m.Position.Y > y {
		m.Position.Y = y
		change = true
	}

	if change {
		m.setRandomParams()
	}
}

func (m *Medicine) calcSpeed() {
	cosX := math.Cos(m.Angle)
	sinY := math.Sin(m.Angle)

	m.vectorSpeed.Set(
		m.Speed*cosX,
		m.Speed*sinY,
	)

	m.dirty = false
}

func (m *Medicine) ToData() *MedicineData {
	m.Data.Id = m.Id
	m.Data.Type = ENTITY_TYPE_MEDICINE
	m.Data.X = math.Floor(m.Position.X)
	m.Data.Y = math.Floor(m.Position.Y)
	m.Data.Radius = math.Floor(m.Radius)
	m.Data.Angle = math.Floor(m.Angle * utils.RAD_TO_DEG)
	return m.Data
}

type MedicineData struct {
	Id                  int
	Type                EntityType
	Angle, X, Y, Radius float64
}

func (d *MedicineData) GetCreateBuffer() []byte {
	x := int(d.X)
	y := int(d.Y)
	r := int(d.Radius)
	a := int(d.Angle)

	return []byte{
		byte(d.Id >> 8 & 0xff),
		byte(d.Id >> 0 & 0xff),

		byte(ENTITY_TYPE_MEDICINE),

		byte(a >> 8 & 0xff),
		byte(a >> 0 & 0xff),

		byte(x >> 8 & 0xff),
		byte(x >> 0 & 0xff),

		byte(y >> 8 & 0xff),
		byte(y >> 0 & 0xff),

		byte(r >> 8 & 0xff),
		byte(r >> 0 & 0xff),
	}
}

func (d *MedicineData) GetUpdateBuffer() []byte {
	x := int(d.X)
	y := int(d.Y)
	a := int(d.Angle)

	return []byte{
		byte(d.Id >> 8 & 0xff),
		byte(d.Id >> 0 & 0xff),

		byte(a >> 8 & 0xff),
		byte(a >> 0 & 0xff),

		byte(x >> 8 & 0xff),
		byte(x >> 0 & 0xff),

		byte(y >> 8 & 0xff),
		byte(y >> 0 & 0xff),
	}
}

func (d *MedicineData) Copy() MedicineData {
	return MedicineData{
		Id:     d.Id,
		Type:   d.Type,
		Angle:  d.Angle,
		X:      d.X,
		Y:      d.Y,
		Radius: d.Radius,
	}
}

func (d *MedicineData) ShouldUpdate(d2 *MedicineData) bool {
	return d.X != d2.X || d.Y != d2.Y || d.Angle != d2.Angle
}

File: src/server/Medicine.ts
/**
 * Created by nazarigonzalez on 15/7/16.
 */
import Point from "./Point";
import {MEDICINE_DATA, ENTITY_TYPE} from "../common/interfaces";
import {Actor} from "./Actor";
import * as rbush from "rbush";
import ObjPool from "obj-pool";
import {arrayPool} from "../common/utils";

enum STATE {
  BORN, ALIVE, DEAD
}

const RADAR_TIME:number = 1;
const RAD_TO_DEG:number = 180/Math.PI;
const RAD:number = 2*Math.PI;
const rad180:number = Math.PI;
const RADIUS:number = 75;
export const MEDICINE_SEARCH_DISTANCE:number = 0.25;
const TIME_TO_RESPAWN:number = 90;
const SPEED_MIN = 50;
const SPEED_MAX = 130;

export class Medicine {
  static pool:ObjPool<any> = new ObjPool(Medicine);
  static STATE = STATE;

  private _state:STATE = STATE.BORN;
  private _speedMin:number = SPEED_MIN;
  private _speedMax:number = SPEED_MAX;
  private _speed:number = 0;
  private _angle:number = 0;

  private _vectorSpeed:Point = new Point();
  private _changeMovementTimer:number = 0;
  private _target:Actor;
  private _targetRadarTime:number = 0;
  private _dirty:boolean = false;
  private _distance:number;

  position:Point = new Point();
  radius:number = RADIUS;
  angleDir:number = 0;
  box:rbush.ItemInterface = {minX: 0, minY: 0, maxX: 0, maxY: 0, id:-1};
  id:number;
  mapSize:number;

  timeToRespawn:number = 0;
  searchTime:number = 0;

  constructor(){}

  initialize(id:number, mapSize:number){
    this.id = id;
    this.mapSize = mapSize;
    this._distance = mapSize*MEDICINE_SEARCH_DISTANCE;
    this._speedMax = SPEED_MAX + (Math.random()*30);
    this.setRandomParams();

    this.box["id"] = this.id;
  }

  reset(){
    this.id = -1;
    this.mapSize = 0;
    this.radius = RADIUS;
    this._speed = 0;
    this._angle = 0;
    this.angleDir = 0;
    if(this._target){
      this._target.findingByMedicine = -1;
    }
    this._target = null;
    this._vectorSpeed.set(0);
    this._changeMovementTimer = 0;
    this._targetRadarTime = 0;
    this._dirty = false;
    this._distance = 0;
    this.position.set(0);
    this.timeToRespawn = 0;
    this.searchTime = 0;
  }

  die(){
    this.state = STATE.DEAD;
    this.timeToRespawn = TIME_TO_RESPAWN + Math.random()*30;
  }

  update(delta:number){
    if(this.state === STATE.DEAD){
      this.timeToRespawn -= delta;
      return;
    }

    this.position.x += this._vectorSpeed.x*delta;
    this.position.y += this._vectorSpeed.y*delta;

    if(this._target){

      if(this._target.state === Actor.STATE.DEAD || this._target.crazy){
        //console.log('LOST TARGET', this._target.name);
        this._target.findingByMedicine = -1;
        this._target = null;
        this.setRandomParams();
      }else{

        this._targetRadarTime -= delta;
        if(this._targetRadarTime <= 0){
          const a:number = this.position.x - this._target.position.x;
          const b:number = this.position.y - this._target.position.y;
          const distance:number = Math.abs(Math.sqrt(a*a + b*b));

          if(distance > this._distance + this._target.radius){
            //console.log('LOST TARGET', this._target.name);
            this._target.findingByMedicine = -1;
            this._target = null;
            this.setRandomParams();
          }else{
            this.angleDir = Math.atan2(this._target.position.y - this.position.y, this._target.position.x - this.position.x)%RAD;
          }
        }

      }
    }else{
      this._changeMovementTimer -= delta;
      if(this._changeMovementTimer <= 0){
        this.setRandomParams();
      }
    }

    if(this.angleDir !== this._angle){
      if(this._angle < this.angleDir){
        this._angle = (this._angle + rad180*delta)%RAD;
        if(this._angle > this.angleDir)this._angle=this.angleDir;
      }else{
        this._angle = (this._angle - rad180*delta)%RAD;
        if(this._angle < this.angleDir)this._angle=this.angleDir;
      }
      this._dirty = true;
    }

    if(this._dirty)this.__calcSpeed();
  }

  updateBox(){
    this.box.minX = this.position.x - this.radius;
    this.box.minY = this.position.y - this.radius;
    this.box.maxX = this.position.x + this.radius;
    this.box.maxY = this.position.y + this.radius;
  }

  setRandomParams(){
    if(this._state === STATE.BORN)this._state = STATE.ALIVE;
    this.angleDir = (Math.floor(Math.random()*360) * Math.PI/180)%RAD;
    this._speed = this._speedMin;
    this._changeMovementTimer = 3 + Math.random() * 6; //seconds
    this.__calcSpeed();
  }

  setTarget(target:Actor){
    if(this._target){
      this._target.findingByMedicine = -1;
    }

    this._targetRadarTime = RADAR_TIME;
    this._target = target;
    this._target.findingByMedicine = this.id;
    this._speed = this._speedMax;
    this.angleDir = Math.atan2(this._target.position.y - this.position.y, this._target.position.x - this.position.x)%RAD;
    this.__calcSpeed();
    //console.log('## Search target', this._target.name);
  }

  isBestTarget(target:Actor):boolean {
    if(!this._target||this._target.state === Actor.STATE.DEAD||target.crazy)return true;
    return target.mass > this._target.mass;
  }

  fixPosition(delta:number, xBound:number, yBound:number){
    let change:boolean = false;

    if(this.position.x < -xBound){
      this.position.x = -xBound;
      change = true;
    }else if(this.position.x > xBound){
      this.position.x = xBound;
      change = true;
    }

    if(this.position.y < -yBound){
      this.position.y = -yBound;
      change = true;
    }else if(this.position.y > xBound){
      this.position.y = yBound;
      change = true;
    }

    if(change){
      this.setRandomParams();
    }
  }

  toArray():MEDICINE_DATA{
    let arr:any[] = arrayPool.alloc();
    arr[0] = this.id;
    arr[1] = ENTITY_TYPE.MEDICINE;
    arr[2] = Math.round(this._angle*RAD_TO_DEG);
    arr[3] = Math.round(this.position.x);
    arr[4] = Math.round(this.position.y);
    arr[5] = Math.round(this.radius);
    return arr as MEDICINE_DATA;
  }

  private __calcSpeed(){
    const cosX:number = Math.cos(this._angle);
    const sinY:number = Math.sin(this._angle);

    this._vectorSpeed.set(
      this._speed * cosX,
      this._speed * sinY
    );

    this._dirty = false;
  }

  get state():STATE {return this._state;};
  set state(value:STATE){
    if(value !== this._state){
      this._state = value;
    }
  }

  get target():Actor{return this._target;};
}
